yarn-project/simulator/src/avm/serialization/bytecode_serialization.ts

```diff
@@ -179,18 +180,27 @@ export function decodeInstructionFromBytecode(
   instructionSet: InstructionSet = INSTRUCTION_SET(),
 ): [Instruction, number] {
   if (pc >= bytecode.length) {
-    throw new Error(`pc ${pc} is out of bounds for bytecode of length ${bytecode.length}`);
+    throw new InvalidProgramCounterError(pc, bytecode.length);
   }
-  const cursor = new BufferCursor(bytecode, pc);
-  const startingPosition = cursor.position();
-  const opcode: Opcode = cursor.bufferAtPosition().readUint8(); // peek.
 
-  const instructionDeserializerOrUndef = instructionSet.get(opcode);
-  if (instructionDeserializerOrUndef === undefined) {
-    throw new Error(`Opcode ${Opcode[opcode]} (0x${opcode.toString(16)}) not implemented`);
+  try {
+    const cursor = new BufferCursor(bytecode, pc);
+    const startingPosition = cursor.position();
+    const opcode: Opcode = cursor.bufferAtPosition().readUint8(); // peek.
```

Comment:
Related to my previous comment, I wonder why this works type-wise. Only some uint8's are opcodes.

Reply:
Yep, reading that unfortunately numeric (default) enums are not type safe. They should merely be treated as a set of constants. Apparently, defining them with string values is more robust but it is not really what we want.